internal/set.go

```diff
@@ -0,0 +1,121 @@
+package internal
+
+import (
+	"encoding/binary"
+	"fmt"
+	"runtime"
+	"sync/atomic"
+)
+
+const unknownCompileError = "unknown error compiling pattern"
+
+type Set struct {
+	ptr      wasmPtr
+	abi      *libre2ABI
+	opts     CompileOptions
+	exprs    []string
+	released uint32
+}
+
+func CompileSet(exprs []string, opts CompileOptions) (*Set, error) {
+	abi := newABI()
+	setPtr := newSet(abi, opts)
+	set := &Set{
+		ptr:   setPtr,
+		abi:   abi,
+		opts:  opts,
+		exprs: exprs,
+	}
+	var estimatedMemorySize int
+	for _, expr := range exprs {
+		estimatedMemorySize += len(expr) + 2
+	}
+
+	alloc := abi.startOperation(estimatedMemorySize)
+	defer abi.endOperation(alloc)
+
+	for _, expr := range exprs {
+		cs := alloc.newCString(expr)
+		errMsg := setAdd(set, cs)
+		if errMsg != "" {
+			return nil, fmt.Errorf("%s", errMsg)
+		}
+	}
+	setCompile(set)
+	// Use func(interface{}) form for nottinygc compatibility.
+	runtime.SetFinalizer(set, func(obj interface{}) {
+		obj.(*Set).release()
+	})
+	return set, nil
+}
+
+func (set *Set) release() {
+	if !atomic.CompareAndSwapUint32(&set.released, 0, 1) {
+		return
+	}
+	deleteSet(set.abi, set.ptr)
+}
+
+// Find searches for the first occurrence of any pattern in the Set within the given byte slice.
+// It returns the index of the matched pattern or -1 if no match is found.
+//
+// Parameters:
+// - b: The byte slice to search within.
+//
+// Returns:
+// - int: The index of the matched pattern, or -1 if no match is found.
+func (set *Set) Find(b []byte) int {
```

Comment:
Sorry if my request wasn't clear. Please remove this method and instead add `FindAllString(s string, n int) []int`. It is somewhat similar to

https://github.com/wasilibs/go-re2/blob/main/internal/re2.go#L340

Also note how there we define a private `findAll` method that is shared regardless of if the input is `[]byte` of `string`, this one will be simpler but we should use the similar pattern.